src/transformers/models/pop2piano/tokenization_pop2piano.py

```diff
@@ -0,0 +1,501 @@
+
+    return current_idx
+
+
+def token_note_to_note(number, current_velocity, default_velocity, note_onsets_ready, current_idx, notes):
+    if note_onsets_ready[number] is not None:
+        # offset with onset
+        onset_idx = note_onsets_ready[number]
+        if onset_idx < current_idx:
+            # Time shift after previous note_on
+            offset_idx = current_idx
+            notes.append([onset_idx, offset_idx, number, default_velocity])
+            onsets_ready = None if current_velocity == 0 else current_idx
+            note_onsets_ready[number] = onsets_ready
+    else:
+        note_onsets_ready[number] = current_idx
+    return notes
+
+
+class Pop2PianoTokenizer(PreTrainedTokenizer):
+    """
+    Constructs a Pop2Piano tokenizer. This tokenizer does not require training.
+
+    This tokenizer inherits from [`PreTrainedTokenizer`] which contains most of the main methods. Users should refer
+    to: this superclass for more information regarding those methods. However the code does not allow that and only
+    supports composing from various genres.
+
+    Args:
+        vocab_file (`str`):
+            Path to the tokenizer file which contains token-ids such as `TOKEN_SPECIAL`, `DEFAULT_VELOCITY`.
+    """
+
+    vocab_files_names = VOCAB_FILES_NAMES
+    pretrained_vocab_files_map = PRETRAINED_VOCAB_FILES_MAP
+
+    def __init__(
+        self,
+        vocab_file,
+        unk_token="-1",
+        eos_token="1",
+        pad_token="0",
+        bos_token="2",
+        **kwargs,
+    ):
+        unk_token = AddedToken(unk_token, lstrip=False, rstrip=False) if isinstance(unk_token, str) else unk_token
+        eos_token = AddedToken(eos_token, lstrip=False, rstrip=False) if isinstance(eos_token, str) else eos_token
+        pad_token = AddedToken(pad_token, lstrip=False, rstrip=False) if isinstance(pad_token, str) else pad_token
+        bos_token = AddedToken(bos_token, lstrip=False, rstrip=False) if isinstance(bos_token, str) else bos_token
+
+        super().__init__(
+            unk_token=unk_token,
+            eos_token=eos_token,
+            pad_token=pad_token,
+            bos_token=bos_token,
+            **kwargs,
+        )
+
+        with open(vocab_file, "rb") as t_file:
+            self.encoder = json.load(t_file)
+
+        self.vocab_size_special = self.encoder["vocab_size_special"]
+        self.vocab_size_note = self.encoder["vocab_size_note"]
+        self.vocab_size_velocity = self.encoder["vocab_size_velocity"]
+        self.vocab_size_time = self.encoder["vocab_size_time"]
+        self.num_bars = self.encoder["num_bars"]
+
+    @property
+    def vocab_size(self):
+        """Returns the vocabulary size of the tokenizer."""
+        return self.vocab_size_special + self.vocab_size_note + self.vocab_size_time + self.vocab_size_velocity
+
+    def get_vocab(self):
+        """Returns the vocabulary of the tokenizer."""
+        return self.encoder
```

Comment:
This is a bit misleading as it doesn't return the vocab, it returns a dictionary with information about the tokenizer. 

Much of the tokenizer logic, which this tokenizer inherits from `PreTrainedTokenizer` relies on certain asumptions about how tokens are mapped. At the moment, there is no vocabulary for the model.  

@ArthurZucker how would you suggest this is resolved? 

We could for example, have a series of dictionaries for each of the vocabs, which can be returned as a nested dict here e.g. `vocab_special.json`, `vocab_note.json` are read in to the encoder which has each of the vocabs stored as keys e.g. `encoder['special']`, `encoder[`time`]` etc.

Reply:
In the recent commit I have implemented the vocabulary for this tokenizer, we now have  4 vocab files for 4 different token types "TIME", "NOTE", "VELOCITY" and "SPECIAL" and another file for vocab params, now when the `get_vocab` is called, the tokenizer returns a nested dict of those vocab files. Please let me know this design complies with the library or not.

Reply:
So if I understand it correctly I am going to just create 4 vocabs `self.token_time_vocab`, `self.token_special_vocab`, `self.token_note_vocab` and `self.token_velocity_vocab` and create and return the nesting dict in the only get_vocab.

Reply:
Yes! That should be the most readable